extensions-core/spark-extensions/src/main/scala/org/apache/druid/spark/clients/DruidClient.scala

```diff
@@ -0,0 +1,180 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.spark.clients
+
+import com.fasterxml.jackson.core.`type`.TypeReference
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.google.common.net.HostAndPort
+import org.apache.druid.java.util.common.{ISE, StringUtils}
+import org.apache.druid.java.util.http.client.response.{StringFullResponseHandler,
+  StringFullResponseHolder}
+import org.apache.druid.java.util.http.client.{HttpClient, Request}
+import org.apache.druid.query.Druids
+import org.apache.druid.query.metadata.metadata.{ColumnAnalysis, SegmentAnalysis,
+  SegmentMetadataQuery}
+import org.apache.druid.spark.MAPPER
+import org.apache.druid.spark.configuration.{Configuration, DruidConfigurationKeys}
+import org.apache.druid.spark.mixins.Logging
+import org.jboss.netty.handler.codec.http.{HttpMethod, HttpResponseStatus}
+import org.joda.time.{Duration, Interval}
+
+import java.net.URL
+import java.util.{List => JList}
+import javax.ws.rs.core.MediaType
+import scala.collection.JavaConverters.{asScalaBufferConverter, mapAsJavaMapConverter,
+  mapAsScalaMapConverter, seqAsJavaListConverter}
+
+/**
+  * Going with SegmentMetadataQueries despite the significant overhead because there's no other way
+  * to get accurate information about Druid columns.
+  *
+  * This is pulled pretty directly from the druid-hadoop-input-format project on GitHub. There is
+  * likely substantial room for improvement.
+  *
+  * @param httpClient
+  * @param objectMapper
+  * @param hostAndPort
+  */
+class DruidClient(
+                   val httpClient: HttpClient,
+                   val objectMapper: ObjectMapper,
+                   val hostAndPort: HostAndPort
+                 ) extends Logging {
+  private val RETRY_COUNT = 5
+  private val RETRY_WAIT_SECONDS = 5
+  private val TIMEOUT_MILLISECONDS = 300000
+  private val druidBaseQueryURL: HostAndPort => String =
+    (hostAndPort: HostAndPort) => s"http://$hostAndPort/druid/v2/"
+
+  /**
+    * The SQL system catalog tables are incorrect for multivalue columns and don't have accurate
+    * type info, so we need to fall back to segmentMetadataQueries. Given a DATASOURCE and a range
+    * of INTERVALS to query over, return a map from column name to a tuple of
+    * (columnType, hasMultipleValues). Note that this is a very expensive operation over large
+    * numbers of segments. If possible, this method should only be called with the most granular
+    * starting and ending intervals instead of over a larger interval.
+    *
+    * @param dataSource The Druid dataSource to fetch the schema for.
+    * @param intervals  The intervals to return the schema for.
+    * @return A map from column name to data type and whether or not the column is multi-value
+    *         for the schema of DATASOURCE.
+    */
+  def getSchema(dataSource: String, intervals: List[Interval]): Map[String, (String, Boolean)] = {
```

Comment:
Should intervals be optional? If interval is not provided, the default interval specified by the broker config `druid.query.segmentMetadata.defaultHistory` will be used. https://druid.apache.org/docs/latest/querying/segmentmetadataquery.html#intervals

Reply:
Intervals are required because the default behavior is unlikely to be what Spark users would expect (e.g. if no time bounds are provided, I would expect to read the whole table in a batch context, not just the last week.) For now, this is only called in [one place](https://github.com/apache/druid/pull/10920/files/331e0c9d088cad3778f3a57432680a1cc607c084#diff-fdc08076d81ecaaf0431bfa58ec51cd8ea8c9efd6aa2ba72faf0eea9928c0ffaR76), which uses `JodaTime.MIN_INSTANT` and `JodaTime.MAX_INSTANT` as the interval bounds if no time filters are provided. This will be an expensive query for large datasources, but until #9707 is implemented I don't know of a better way. I suppose it would make sense to move that default handling into the `getSchema` method here instead though, so I'll make the change.